### PR TITLE
CDAP-4270: CLI for metadata and lineage

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-4270
http://builds.cask.co/browse/CDAP-DUT3296

### CODE REVIEW

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import co.cask.cdap.cli.util.AbstractCommand;
+import co.cask.cdap.client.LineageClient;
+import co.cask.cdap.proto.Id;
+import co.cask.common.cli.Arguments;
+import com.google.inject.Inject;
+
+import java.io.PrintStream;
+
+/**
+ * Call an endpoint of a {@link Service}.
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java]: update javadocs

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import co.cask.cdap.cli.util.AbstractCommand;
+import co.cask.cdap.client.LineageClient;
+import co.cask.cdap.proto.Id;
+import co.cask.common.cli.Arguments;
+import com.google.inject.Inject;
+
+import java.io.PrintStream;
+
+/**
+ * Call an endpoint of a {@link Service}.
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java]: update javadocs

File: cdap-cli/src/main/java/co/cask/cdap/cli/commandset/MetadataCommands.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import co.cask.cdap.cli.command.metadata.RemoveMetadataTagCommand;
+import co.cask.cdap.cli.command.metadata.RemoveMetadataTagsCommand;
+import co.cask.common.cli.Command;
+import co.cask.common.cli.CommandSet;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+
+/**
+ * Application commands.
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/commandset/MetadataCommands.java]: update javadocs

File: cdap-cli/src/main/java/co/cask/cdap/cli/commandset/MetadataCommands.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        .add(injector.getInstance(RemoveMetadataPropertiesCommand.class))
+        .add(injector.getInstance(RemoveMetadataPropertyCommand.class))
+        .add(injector.getInstance(RemoveMetadataTagCommand.class))
+        .add(injector.getInstance(RemoveMetadataTagsCommand.class))
+        .build());
+  }
+
+  @Override
+  public String getCategory() {
+    return CommandCategory.APPLICATION_LIFECYCLE.getName();
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/commandset/MetadataCommands.java]: update category

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    long start = getTimestamp(arguments.getOptional("start", "min"), currentTime);
+    long end = getTimestamp(arguments.getOptional("end", "max"), currentTime);
+    Integer levels = arguments.getIntOptional("levels", null);
+
+    client.getLineage(dataset, start, end, levels);
+  }
+
+  @Override
+  public String getPattern() {
+    return String.format("get lineage dataset <%s>", ArgumentName.DATASET);
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java]: the pattern doesn't reflect start and end and levels?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (rohitsinha54) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java]: 2015

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java]: 2015

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  @Override
+  public void perform(Arguments arguments, PrintStream output) throws Exception {
+    long currentTime = System.currentTimeMillis();
+    Id.DatasetInstance dataset = Id.DatasetInstance.from(
+      cliConfig.getCurrentNamespace(), arguments.get(ArgumentName.DATASET.toString()));
+    long start = getTimestamp(arguments.getOptional("start", "min"), currentTime);
+    long end = getTimestamp(arguments.getOptional("end", "max"), currentTime);
+    Integer levels = arguments.getIntOptional("levels", null);
+
+    client.getLineage(dataset, start, end, levels);
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetDatasetLineageCommand.java]: need to print the output to console somehow.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Override
+  public void perform(Arguments arguments, PrintStream output) throws Exception {
+    long currentTime = System.currentTimeMillis();
+    Id.Stream stream = Id.Stream.from(cliConfig.getCurrentNamespace(), arguments.get(ArgumentName.STREAM.toString()));
+    long start = getTimestamp(arguments.getOptional("start", "min"), currentTime);
+    long end = getTimestamp(arguments.getOptional("end", "max"), currentTime);
+    Integer levels = arguments.getIntOptional("levels", null);
+
+    client.getLineage(stream, start, end, levels);
```
Comment (anwar6953) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/lineage/GetStreamLineageCommand.java]: need to print the output to console somehow.

**Comment by anwar6953** [comment]
image

This is what the current rendering looks like.
Since its only one row, maybe it could be improved (for instance, not rendered as a table - or maybe the relations be rendered in rows of a table?)
@poornachandra what do you think?

**Comment by poornachandra** [comment]
We could do simple ascii rendering of the DAG.

**Comment by anwar6953** [comment]
An alternative to the current rendering is json formatting (like `python -mjson.tool` does)

**Comment by anwar6953** [comment]
image

This is definitely an improvement; thanks.

**Comment by jwang47** [comment]
@poornachandra Created https://issues.cask.co/browse/CDAP-4395

**Comment by anwar6953** [comment]
LGTM

**Pull request merged**